src/mono/wasm/runtime/hybrid-globalization/change-case.ts

```diff
@@ -13,15 +13,37 @@ export function mono_wasm_change_case_invariant(src: number, srcLength: number,
     const exceptionRoot = mono_wasm_new_external_root<MonoObject>(ex_address);
     try{
         const input = get_utf16_string(src, srcLength);
-        let result = toUpper ? input.toUpperCase() : input.toLowerCase();
+        const result = toUpper ? input.toUpperCase() : input.toLowerCase();
+
         // Unicode defines some codepoints which expand into multiple codepoints,
         // originally we do not support this expansion
-        if (result.length > dstLength)
-            result = input;
+        if (result.length <= dstLength)
+        {
+            for (let i = 0; i < result.length; i++)
+                setU16(dst + i*2, result.charCodeAt(i));
+            wrap_no_error_root(is_exception, exceptionRoot);
+            return;
+        }
 
-        for (let i = 0; i < result.length; i++)
-            setU16(dst + i*2, result.charCodeAt(i));
-        wrap_no_error_root(is_exception, exceptionRoot);
+        // workaround to maintain the ICU-like behavior
+        if (toUpper)
+        {
+            for (let i=0; i < input.length; i++)
+            {
+                const upperChar = input[i].toUpperCase();
```

Comment:
Are there any multi surrogate characters which would break with this operation ?

Reply:
Yes, lower "\uD801\uDC37" -> upper "\uD801\uDC0F"